subtests/docker_cli/run_env/run_env.py

```diff
@@ -3,493 +3,123 @@
-        Check that variable's value matches the exp regexp. Fails the test
-        when it's not.
-        :param exp: Expected value (string)
-        :param variable: Name of the env variable (string)
-        :param env: Environment (dictionary)
-        :param container: Name of the container (string)
-        :param appendix: Rest of the error string
-        """
-        self.failif(not re.match(exp, env.get(variable)), "%s=%s not found in"
-                    " %s env:\n%s%s" % (variable, exp, container, env,
-                                        appendix))
-
-    def init_container(self, prefix, subargs, cmd):
-        """
-        Prepares dkrcmd and stores the name in self.sub_stuff['containers']
-        :return: tuple(dkrcmd, name)
-        """
-        name = self.sub_stuff['dc'].get_unique_name(prefix)
-        subargs.append("--name")
-        subargs.append(name)
-        self.logdebug("Queuing container %s for removal", name)
-        self.sub_stuff['containers'].append(name)
-        fin = DockerImage.full_name_from_defaults(self.config)
-        subargs.append(fin)
-        subargs.append(cmd)
-        dkrcmd = InteractiveAsyncDockerCmd(self, 'run', subargs)
-        dkrcmd.verbose = True
-        dkrcmd.quiet = False
-        return dkrcmd, name
-
-    def initialize(self):
-        super(run_env_base, self).initialize()
-        # Prepare a container
-        config.none_if_empty(self.config)
-        self.sub_stuff['dc'] = DockerContainers(self)
-        self.sub_stuff['containers'] = []
-
-    def cleanup(self):
-        super(run_env_base, self).cleanup()
-        if self.config['remove_after_test']:
-            dc = DockerContainers(self)
-            dc.clean_all(self.sub_stuff.get("containers", []))
-
-
-class port_base(run_env_base):
-
-    """
-    Base class for port test.
-    """
+    pass
+
+
+class spam(SubSubtest):
+
+    # Allows quickly rejecting container env. supplied vars
+    KEYPFX = '_x'
+    VALPFX = 'x_'
+
+    @classmethod
+    def make_test_dict(cls, howmany, keyiter=None, valiter=None):
+        xcept = ValueError("Must specify neither or both "
+                           "keyiter and valiter")
+        if ((keyiter is not None and valiter is None) or
+                (keyiter is None and valiter is not None)):
+            raise xcept
+        keyfmt = "%s%%x" % cls.KEYPFX
+        valfmt = "%s%%x" % cls.VALPFX
+        # Because of above, only need to check one
+        if keyiter is None:
+            keyiter = (keyfmt % x for x in xrange(0, howmany))
+            valiter = (valfmt % x for x in xrange(howmany, 0, -1))
+        return dict(zip(keyiter, valiter))
+
+    def save_testdata(self):
+        self.sub_stuff['output'] = os.path.join(self.tmpdir, 'env.output')
+        resultdir = self.parent_subtest.resultsdir
+        inputcopy = os.path.join(resultdir, "env.input")
+        with open(inputcopy, 'wb') as inpf:
+            for key, value in self.sub_stuff['envd'].iteritems():
```

Comment:
One of my habits is to make heavy use of sorting. It makes no difference whatsoever to the tool, but when things go wrong the human debugger will appreciate being able to 'diff env.input env.output', or just scan through them, or perhaps even compare against previous runs. Minor nit, and unlikely ever to be needed, but unlikely != impossible.